libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp

```diff
@@ -1002,8 +998,372 @@ bool AP_Frsky_Telem::get_telem_data(uint8_t &frame, uint16_t &appid, uint32_t &d
     return singleton->_get_telem_data(frame, appid, data);
 }
 
-namespace AP {
-    AP_Frsky_Telem *frsky_telem() {
-        return AP_Frsky_Telem::get_singleton();
+#if HAL_WITH_FRSKY_TELEM_BIDIRECTIONAL
+/*
+  fetch Sport data for an external transport, such as FPort
+ */
+bool AP_Frsky_Telem::_set_telem_data(const uint8_t frame, const uint16_t appid, const uint32_t data)
+{
+    sport_packet_t sp{};
+
+    sp.sensor = 0x0D;
+    sp.frame = frame;
+    sp.appid = appid;
+    sp.data = data;
+
+    // queue only Uplink packets
+    if (sp.frame == SPORT_UPLINK_FRAME || sp.frame == SPORT_UPLINK_FRAME_RW) {
+        _sport_rx_packet_buffer.push_force(sp);
+        return true;
+    }
+    return false;
+}
+
+/*
+  fetch Sport data from an external transport, such as FPort
+ */
+bool AP_Frsky_Telem::set_telem_data(const uint8_t frame, const uint16_t appid, const uint32_t data)
+{
+    if (!singleton && !hal.util->get_soft_armed()) {
+        // if telem data is requested when we are disarmed and don't
+        // yet have a AP_Frsky_Telem object then try to allocate one
+        new AP_Frsky_Telem(true);
+        if (singleton) {
+            singleton->init();
+        }
+    }
+    if (!singleton) {
+        return false;
+    }
+    return singleton->_set_telem_data(frame, appid, data);
+}
+
+
+/*
+ * Queue uplink packets in the sport rx queue
+ * for FrSky SPort Passthrough (OpenTX) protocol (X-receivers)
+ */
+void AP_Frsky_Telem::process_sport_telemetry_packet(const uint8_t * packet)
+{
+    if (!check_sport_packet(packet, _last_sport_packet)) {
+        return;
+    }
+
+    sport_packet_t sp{};
+
+    sp.sensor = packet[0];
+    sp.frame = packet[1];
+    sp.appid = le16toh_ptr(packet+2);
+    sp.data = le32toh_ptr(packet+4);
+
+    // queue only Uplink packets
+    if (sp.sensor == _sport_config.uplink_sensor_id && sp.frame == SPORT_UPLINK_FRAME) {
+        _sport_rx_packet_buffer.push_force(sp);
+    }
+}
+
+/*
+ * Process the sport rx queue
+ * Extract up to 1 mavlite message from the queue
+ * for FrSky SPort Passthrough (OpenTX) protocol (X-receivers)
+ */
+void AP_Frsky_Telem::process_sport_rx_queue()
+{
+    bool empty = _sport_rx_packet_buffer.is_empty();
+
+    while (!empty) {
+        sport_packet_t packet{};
+        {
+            _sport_rx_packet_buffer.peek(packet);
+            for (uint8_t i=0; i<6; i++) {
+                mavlite_rx_parse(packet.raw[i+2],i, &_mavlite_rx_message, &_mavlite_rx_status);
+            }
+
+            _sport_rx_packet_buffer.pop();
+            empty = _sport_rx_packet_buffer.is_empty();
+        }
+
+        if (_mavlite_rx_status.parse_state == PARSE_STATE_MESSAGE_RECEIVED) {
```

Comment:
I'll dive into the mavlink-lite parser next review - but I'm curious about the six-byte boundary thing going on, and whether we could make `mavlink_rx_parse` non-void indicating it has parsed a packet.

Reply:
we can for sure, I did the same change in the frsky packet parser, I'll do it